runtime/src/bank.rs

```diff
@@ -3178,6 +3165,20 @@ impl Bank {
         consumed_budget.saturating_sub(budget_recovery_delta)
     }
 
+    // This is called from snapshot restore and for each epoch boundary
+    // The entire code path herein must be idempotent
+    pub fn refresh_programs_and_inflation(&mut self) {
+        if let Some(entered_epoch_callback) =
+            self.entered_epoch_callback.clone().read().unwrap().as_ref()
+        {
+            entered_epoch_callback(self)
+        }
+
+        for program in get_builtins(self.operating_mode(), self.epoch()) {
+            self.add_builtin(&program.name, program.id, program.entrypoint);
```

Comment:
Adding the same programs at every epoch is going to fail because the account already exists.  Looks like the bank hash ci failure is related, looking into it.